third_party/move/move-vm/runtime/src/runtime_type_checks_async.rs

```diff
@@ -0,0 +1,785 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+//! Defines async type checker that abstractly interprets Move bytecode to perform type checks
+//! based on the execution trace.
+//!
+//! The type checker should ideally be run in parallel and as a post-execution hook. Otherwise,
+//! if there is no parallelism or not enough transactions, running checks in-place is preferred.
+//!
```

Comment:
I'm not sure we should keep two ways. Or at least if we do, we need to avoid duplication of code. It appears currently you have the old check mechanism unchanged?

You can use that `TraceLogger` interface to connect to a trace based verifier like this here, provided you have designed the APIs in the right way, avoiding the construction of the intermediate data.